core/src/main/java/org/elasticsearch/cluster/routing/UnassignedInfo.java

```diff
@@ -126,6 +129,9 @@ private UnassignedInfo(Reason reason, long timestamp, String message, Throwable
     UnassignedInfo(StreamInput in) throws IOException {
         this.reason = Reason.values()[(int) in.readByte()];
         this.timestamp = in.readLong();
+        // As System.nanoTime() cannot be compared across different JVMs, reset it to now.
+        // This means that in master failover situations, elapsed delay time is forgotten.
```

Comment:
What this does is reduce the delay of a delayed allocation by assuming that clocks on old and new master node are approx. in sync. What if they are not? Shard gets assigned too early. Not sure about tradeoffs here so I chose the more deterministic approach.